### PR TITLE
Fix/ota demo http

### PR DESCRIPTION
**Pull request by pvyawaha**
*Issue #, if available:*

*Description of changes:*

* Speed up OTA over HTTP demo by checking connection close flag and connecting before sending request 

* Fix issues where buffer was directly used instead of calling get buffer function.

* Added handling of expired url.

* Added exit function to both OTA over MQTT and HTT demos

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 1]
```diff
@@ -712,32 +729,17 @@ static void otaAppCallback( OtaJobEvent_t event,
...
 
         /* We should never get here as new image activation must reset the device.*/
-        LogError( ( "New image activation failed." ) );
+        LogError( ( "New image activation failed, exiting." ) );
 
-        for( ; ; )
-        {
-        }
+        /* Exit this application.*/
+        exit( EXIT_SUCCESS );
```
Comment (yanjos-dev) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:742]: If we failed to activate the image, EXIT_SUCCESS seems to not make sense. Is there a different error code that we can use?

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 2]
```diff
@@ -1178,23 +1180,34 @@ static int establishConnection( void )
...
+        if( pthread_mutex_lock( &mqttMutex ) == 0 )
+        {
+            /* Disconnect MQTT session. */
+            MQTT_Disconnect( &mqttContext );
 
-        pthread_mutex_unlock( &mqttMutex );
+            pthread_mutex_unlock( &mqttMutex );
+
+            /* Clear the mqtt session flag. */
+            mqttSessionEstablished = false;
```
Comment (yanjos-dev) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:1199]: May make more sense to have this inside of the mutex after the disconnect call.

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 3]
```diff
@@ -1270,6 +1283,63 @@ static int32_t connectToS3Server( NetworkContext_t * pNetworkContext,
...
+                /* Send job document received event. */
+                eventMsg.eventId = OtaAgentEventReceivedFileBlock;
+                eventMsg.pEventData = pData;
+                OTA_SignalEvent( &eventMsg );
+
+                ret = OtaHttpSuccess;
+            }
+            else
+            {
+                LogError( ( "Error: No OTA data buffers available.\r\n" ) );
```
Comment (yanjos-dev) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:1315]: \r\n are automatically added to the print statements so they aren't needed here. Unless you are intending to create extra spacing, in that case it's okay.

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 1 above)
Comment (pvyawaha) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c:742]: At this point the device has successfully downloaded the new version, activation only failed as user will have to execute the application manually. So this is not a error , we just log that activation failed.

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 3 above)
Comment (pvyawaha) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1451/commits/463ab1e0eb8ae36da14dfb76e6ab62d316ff0ab1

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 2 above)
Comment (pvyawaha) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c:1202]: Addressed in 
https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1451/commits/463ab1e0eb8ae36da14dfb76e6ab62d316ff0ab1

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 4]
```diff
@@ -204,20 +204,18 @@
  */
 #define CONNECTION_RETRY_MAX_ATTEMPTS    ( 5U )
 
-/* Check that size of the user buffer is defined. */
-#ifndef USER_BUFFER_LENGTH
-    #define USER_BUFFER_LENGTH    ( 2048 )
-#endif
+/* HTTP buffers used for http request and response. */
+#define HTTP_USER_BUFFER_LENGTH          ( otaconfigFILE_BLOCK_SIZE + 1024 )
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1451/commits/463ab1e0eb8ae36da14dfb76e6ab62d316ff0ab1

**Review by yanjos-dev — APPROVED** [review]

**Review by divekarshubham — APPROVED** [review]

**Review by yanjos-dev — APPROVED** [review]

**Pull request closed without merging**